sdk/core/core/src/az_http_private.h

```diff
@@ -27,19 +27,14 @@
 AZ_NODISCARD AZ_INLINE az_result _az_http_request_mark_retry_headers_start(_az_http_request* p_hrb)
 {
   AZ_PRECONDITION_NOT_NULL(p_hrb);
-  p_hrb->_internal.retry_headers_start_byte_offset = az_span_length(p_hrb->_internal.headers);
+  p_hrb->_internal.retry_headers_start_byte_offset = p_hrb->_internal.headers_length * (int32_t)sizeof(az_pair);
   return AZ_OK;
 }
 
 AZ_NODISCARD AZ_INLINE az_result _az_http_request_remove_retry_headers(_az_http_request* p_hrb)
 {
   AZ_PRECONDITION_NOT_NULL(p_hrb);
-
-  az_span* headers_ptr = &p_hrb->_internal.headers;
-  *headers_ptr = az_span_init(
-      az_span_ptr(*headers_ptr),
-      p_hrb->_internal.retry_headers_start_byte_offset,
-      az_span_capacity(*headers_ptr));
+  p_hrb->_internal.headers_length = p_hrb->_internal.retry_headers_start_byte_offset / (int32_t)sizeof(az_pair);
```

Comment:
I don't fully understand what this method is trying to do along with the relationship of `p_hrb->_internal.retry_headers_start_byte_offset` and the `&p_hrb->_internal.headers`. Please review this @antkmsft. I would love to chat offline to make sure I understand what the intention was here.

Reply:
headers is a buffer of bytes, in this case, we use it to hold an array of az_pair.
az_pair is an structure that contains 2 az_span one after the other.

So, instead of having headers as `az_pair[] headers` and request customers to manager building the az_pairs, we decided to use an az_span to hold the az_pair.

So we calling az_append_header, we need to multiply the number of headers * sizeof(az_pair) to know where to copy the az_pair bits.

Then, we have this retry offset as a way separate 2 different types of headers.
All headers added by customer and before reaching the retry policy are considered `pre-retry headers`
While all headers added after and during retry headers are `post-retry headers`

Calling the set_retry_headers_offset would use the current size of headers (number of headers * sizeof(az_pair)) as the beginning of post-retry headers.
Then, if we call clean_retry_headers(..)  we would delete all headers after that offset